# One-Way ANOVA in Python: 
Hypothesis involved:
one-way ANOVA has the below given null and alternative hypotheses:

H0 (null hypothesis): μ1 = μ2 = μ3 = … = μk (It implies that the means of all the population are equal)
H1 (null hypothesis): It states that there will be at least one population mean that differs from the rest

In [9]:


#Scenario:Researchers took 20 cars of the same to take part in a study. These cars are randomly doped with one of 
#the four-engine oils and allowed to run freely for 100 kilometers each. At the end of the journey, the performance of 
#each of the cars is noted
# H0 -> No difefrence in performance for teh cars
# H1 ->There are difference in performance depending on the oil


In [5]:
# Performance when each of the engine
# oil is applied
performance1 = [89, 89, 88, 78, 79]
performance2 = [93, 92, 94, 89, 88]
performance3 = [89, 88, 89, 93, 90]
performance4 = [81, 78, 81, 92, 82]


In [7]:
# Importing library
from scipy.stats import f_oneway
# Conduct the one-way ANOVA
f_oneway(performance1, performance2, performance3, performance4)


F_onewayResult(statistic=4.625000000000002, pvalue=0.016336459839780215)

In [ ]:
 #Since the p-value is less than 0.05 hence we would reject the null hypothesis. 
#This means we have sufficient proof to say that there exists a difference in the performance among four different engine oils. 

# Two-Way ANOVA : 
scientists need to know whether plant growth is affected by fertilizers and watering frequency. They planted exactly 30 plants and allowed them to grow for six months under different conditions for fertilizers and watering frequency. After exactly six months, they recorded the heights of each plant centimeters. 

H0 -> Null hypthesys ->Watering fertilizer & watering frequency has no effect

H1-> Alternate hypothesys - > There is an effect of watring frequency & fertilizer has an effect on plant growth

In [10]:
# Importing libraries
import numpy as np
import pandas as pd
# Create a dataframe
dataframe = pd.DataFrame({'Fertilizer': np.repeat(['daily', 'weekly'], 15),
'Watering': np.repeat(['daily', 'weekly'], 15),
'height': [14, 16, 15, 15, 16, 13, 12, 11, 14,15, 16, 16, 17, 18, 14, 13, 14, 14,14, 15, 16, 16, 17, 18, 14, 13, 14,14, 14, 15]})

In [11]:
dataframe

Fertilizer Watering  height
0       daily    daily      14
1       daily    daily      16
2       daily    daily      15
3       daily    daily      15
4       daily    daily      16
5       daily    daily      13
6       daily    daily      12
7       daily    daily      11
8       daily    daily      14
9       daily    daily      15
10      daily    daily      16
11      daily    daily      16
12      daily    daily      17
13      daily    daily      18
14      daily    daily      14
15     weekly   weekly      13
16     weekly   weekly      14
17     weekly   weekly      14
18     weekly   weekly      14
19     weekly   weekly      15
20     weekly   weekly      16
21     weekly   weekly      16
22     weekly   weekly      17
23     weekly   weekly      18
24     weekly   weekly      14
25     weekly   weekly      13
26     weekly   weekly      14
27     weekly   weekly      14
28     weekly   weekly      14
29     weekly   weekly      15

In [17]:
# Importing libraries
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Create a dataframe
dataframe = pd.DataFrame({'Fertilizer': np.repeat(['daily', 'weekly'], 15),
                          'Watering': np.repeat(['daily', 'weekly'], 15),
                          'height': [14, 16, 15, 15, 16, 13, 12, 11, 14, 
                                     15, 16, 16, 17, 18, 14, 13, 14, 14, 
                                     14, 15, 16, 16, 17, 18, 14, 13, 14, 
                                     14, 14, 15]})


# Performing two-way ANOVA
model = ols('height ~ C(Fertilizer) + C(Watering) +\
C(Fertilizer):C(Watering)',
data=dataframe).fit()
result = sm.stats.anova_lm(model, type=2)

# Print the result
print(result)


                             df     sum_sq   mean_sq         F    PR(>F)
C(Fertilizer)               1.0   0.033333  0.033333  0.012069  0.913305
C(Watering)                 1.0   0.092308  0.092308  0.033422  0.856260
C(Fertilizer):C(Watering)   1.0   0.057692  0.057692  0.020889  0.886118
Residual                   28.0  77.333333  2.761905       NaN       NaN


# Result
C(Fertilizer):C(Watering)  p value 0.886 >significance 0.05 so accepting null hypothesys means these 2 factors has no effect on plant growth